### ISSUE TITLE
[Performance]: InternVL multi image speed is not improved compare to original

### ISSUE THREAD

**Issue by luohao123** [vllm-project/vllm#9483]
### Your current environment

<details>
<summary>The output of `python collect_env.py`</summary>

```text
latest vllm 0.6.1
```

</details>

### Model Input Dumps

tt

### 🐛 Describe the bug

InternVL multi image speed is slower than original

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

**Comment by DarkLight1337** [comment]
Could you elaborate more? What do you mean by the original speed?

**Comment by luohao123** [comment]
Compare with torch, same device, same dtype (float16, V100).

Single image faster about 20%, while multiple image are slower, A100 got same result.

**Comment by DarkLight1337** [comment]
Can you show the scripts you used to measure the performance of HF vs vLLM?

**Comment by luohao123** [comment]
Hi, the test based on internvl 8b model, have u guys tested vllm speed improvement on multiple images? I am not lying. Multiple images actually slower than torch. For some inhouse issue, I didn't got a chance to paste code here, but I think you guys might can be easily replicate the result.

**Comment by DarkLight1337** [comment]
> Hi, the test based on internvl 8b model, have u guys tested vllm speed improvement on multiple images? I am not lying. Multiple images actually slower than torch. For some inhouse issue, I didn't got a chance to paste code here, but I think you guys might can be easily replicate the result.

No, we have not tested the speed for multiple images (benchmarking work for multi-modal models is still in the early stages). Since vLLM was originally designed around language generation, most of vLLM's optimizations don't currently work on the vision encoder part of the model, which may explain the decrease in speed when more images are passed. We are still busy with making multi-modal support feature complete, so it may take a while before we can focus on optimization - any help is welcome!

**Comment by noooop** [comment]
1. Most multimodal encoder are hardcoded to use F.scaled_dot_product_attention，refer to #8898, [intern_vit.py](https://github.com/vllm-project/vllm/blob/855e0e6f97e5ddd5addf042f25c1f11522214569/vllm/model_executor/models/intern_vit.py#L247)

```
        x = F.scaled_dot_product_attention(q, k, v, scale=self.scale)
```

2.  there is still a noticeable gap between F.scaled_dot_product_attention and FlashAttention. refer to #8453

![image](https://github.com/user-attachments/assets/56728a9b-c7d6-4859-b3de-b659b61f3a07)


**This may be the reason why vllm speed decreases significantly as the number of input images increases**


3. vllm flash_attn backend does not support encoder only model currently, refer to [flash_attn.py](https://github.com/vllm-project/vllm/blob/696b01af8fac1819b2409cc0f205c73ef553558c/vllm/attention/backends/flash_attn.py#L568C1-L572C60) 
Therefore, the encoder only model and encoder-decoder model cannot use the fast flash_attn backend for the time being.

>* FlashAttention backend support for encoder/decoder models is left as future work.  refer to #4888

4. Maybe we can manage to merge #9124, not only for the encoder only model, but also to make the encoder-decoder model faster

To use flash_attn backend for encoder only model and encoder-decoder model requires significant modification of ModelInputBuilder, attention_metadata and Runner, which is not a simple matter

Maybe that's why now I choose to use F.scaled_dot_product_attention

**Comment by luohao123** [comment]
Hi, am not expert in accelerate, but as far as I can understand, why encoder-decoder  can not use flashattn?

**Comment by noooop** [comment]
> Hi, am not expert in accelerate, but as far as I can understand, why encoder-decoder can not use flashattn?

simple answer

> vllm flash_attn backend only support decoder only model currently, refer to [[Kernel] Correctly invoke prefill & decode kernels for cross-attention (towards eventual encoder/decoder model support) #4888](https://github.com/vllm-project/vllm/pull/4888) [flash_attn.py](https://github.com/vllm-project/vllm/blob/696b01af8fac1819b2409cc0f205c73ef553558c/vllm/attention/backends/flash_attn.py#L568C1-L572C60)
> 
> > * FlashAttention backend support for encoder/decoder models is left as future work.


Why backend support encoder-decoder so hard?

Currently the only backend xformer that supports encoder-decoder model is still undergoing bugfixes, refer to #9026 (
I took a look at it before, and I guess there are still a lot of bugs that need to be fixed.

> To use flash_attn backend for encoder only model and encoder-decoder model requires significant modification of ModelInputBuilder, attention_metadata and Runner, which is not a simple matter

**Comment by luohao123** [comment]
How about using flashatten2 package, or using torch inside sdpa?

**Comment by noooop** [comment]
vllm uses vllm_flash_attn, which supports PagedAttention. Of course this is not the point

To use flash_attn backend for encoder-decoder models, you need to pass in the corresponding attention_metadata. 

```
            cu_seqlens_q=attn_metadata.seq_start_loc,
            cu_seqlens_k=attn_metadata.seq_start_loc,
            max_seqlen_q=attn_metadata.max_seq_len,
            max_seqlen_k=attn_metadata.max_seq_len,
```

You can use a diryfix, maybe some features are already fixed and look good

When you're ready to fix everything that's broken. The nightmare begins, you need to modify attention_metadata, ModelInputBuilder, and Runner.

**Comment by noooop** [comment]
PTAL #9559 

There are some subtle differences between Multimodal models and encoder-decoder model, I'm not sure it works on Multimodal models.

**Comment by Jeremy-J-J** [comment]
Same problem

**Comment by torinchen** [comment]
> noooop

when diving into the code, i found that internvl2 uses xformer's attention not the naive one, so, the slow speed maybe come from other part.

**Comment by noooop** [comment]
> when diving into the code, i found that internvl2 uses xformer's attention not the naive one, so, the slow speed maybe come from other part.

@torinchen 

https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/intern_vit.py#L271C1-L285C17


The code of vllm is being optimized rapidly, but at least today 20241029 is still using scaled_dot_product_attention

```
        if self.qk_normalization:
            B_, N_, H_, D_ = q.shape
            q = self.q_norm.forward_native(q.flatten(-2,
                                                     -1)).view(B_, N_, H_, D_)
            k = self.k_norm.forward_native(k.flatten(-2,
                                                     -1)).view(B_, N_, H_, D_)
        q = q.transpose(1, 2)
        k = k.transpose(1, 2)
        v = v.transpose(1, 2)


        x = F.scaled_dot_product_attention(q, k, v, scale=self.scale)
        x = x.transpose(1, 2).view(B, N, -1)


        x = self.proj(x)
        return x
```

**Comment by torinchen** [comment]
> > when diving into the code, i found that internvl2 uses xformer's attention not the naive one, so, the slow speed maybe come from other part.
> 
> @torinchen
> 
> https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/intern_vit.py#L271C1-L285C17
> 
> The code of vllm is being optimized rapidly, but at least today 20241029 is still using scaled_dot_product_attention
> 
> ```
>         if self.qk_normalization:
>             B_, N_, H_, D_ = q.shape
>             q = self.q_norm.forward_native(q.flatten(-2,
>                                                      -1)).view(B_, N_, H_, D_)
>             k = self.k_norm.forward_native(k.flatten(-2,
>                                                      -1)).view(B_, N_, H_, D_)
>         q = q.transpose(1, 2)
>         k = k.transpose(1, 2)
>         v = v.transpose(1, 2)
> 
> 
>         x = F.scaled_dot_product_attention(q, k, v, scale=self.scale)
>         x = x.transpose(1, 2).view(B, N, -1)
> 
> 
>         x = self.proj(x)
>         return x
> ```

https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/intern_vit.py#L353C4-L371C76

**Comment by noooop** [comment]
> when diving into the code, i found that internvl2 uses xformer's attention not the naive one, so, the slow speed maybe come from other part.

vllm installs xformers by default, so internvl2 uses xformer's attention. 

Although xformer is slower than flash attention, it is not significant. 

I agree with @torinchen who is right "the slow speed maybe come from other part."

**Comment by noooop** [comment]
Please submit a test code to reproduce this issues. 

I can help locate the problem using profiler.

**Comment by torinchen** [comment]
[test.zip](https://github.com/user-attachments/files/17552824/test.zip)
i test the online and offline mode, the gap is significant (under lmdeploy, the gap is zero btw )

**Comment by noooop** [comment]
Can this problem be issues using InternVL2-8B?

https://huggingface.co/OpenGVLab/InternVL2-8B

**Comment by torinchen** [comment]
yes, my model is just a sft version InternVL2-8B

**Comment by noooop** [comment]
I can't reproduce this issues

code https://github.com/noooop/light-vllm/tree/main/benchmarks/InternVL2



transformers 4.37.2 + flash_attn 2.6.3

use_flash_attn=True
single-image single-round conversation 1.37133834199999
multi-image single-round conversation 3.133497854799998

transformers 4.45.2 + flash_attn 2.6.3
use_flash_attn=True

single-image single-round conversation 1.4907942284
multi-image single-round conversation 3.1399439033000136


transformers 4.45.2 + vllm==v0.6.3.post1 

single-image single-round conversation 1.367961298399996
multi-image single-round conversation 2.787156264600026

**Comment by luohao123** [comment]
Even though I think the time speed up is very limited, can be treated as not fast as expect.

I didn't forgot my precise number before, but from my human feeling, without streaming, the response of vllm is not fast.

**Comment by noooop** [comment]
> Even though I think the time speed up is very limited, can be treated as not fast as expect.

For a single request, flash attn is already very fast.

vllm can only batched multiple requests to increase throughput.

> I didn't forgot my precise number before, but from my human feeling, without streaming, the response of vllm is not fast.

Do you use openai.api_server or offline inference.

**Comment by torinchen** [comment]
> @torinchen @luohao123
> 
> I can't reproduce this issues
> 
> code https://github.com/noooop/light-vllm/tree/main/benchmarks/InternVL2
> 
> image preprocessing time is not included
> 
> transformers 4.37.2 + flash_attn 2.6.3
> 
> use_flash_attn=True single-image single-round conversation 1.37133834199999 multi-image single-round conversation 3.133497854799998
> 
> transformers 4.45.2 + flash_attn 2.6.3 use_flash_attn=True
> 
> single-image single-round conversation 1.4907942284 multi-image single-round conversation 3.1399439033000136
> 
> transformers 4.45.2 + vllm==v0.6.3.post1
> 
> single-image single-round conversation 1.367961298399996 multi-image single-round conversation 2.787156264600026
> 
> I'm not sure if it's related to the slow speed of image preprocessing #9238

try online mode, i saw speed gap there.

**Comment by noooop** [comment]
> try online mode, i saw speed gap there.

Sorry, I'm not very familiar with the webserver part. 

Many issues mentioned that image preprocessing is slow. I think it is more likely to be caused by this problem.
